### PR TITLE
[14.0][IMP] product_tier_validation: Take into account state Many2one

### PR DESCRIPTION
**Pull request by rousseldenis**
Depends on:

- [ ] https://github.com/OCA/server-ux/pull/421

Remove:

- [ ] test-requirements.txt entry

### CODE REVIEW

**Comment by rousseldenis** [comment]
@simahawk

**Comment by hailangvn** [comment]
I am wondering what if we change https://github.com/OCA/product-attribute/blob/14.0/product_state/views/product_template.xml#L134 to L138 as below.
```
-                    name="product_state_id"
+                    name="state"
                     widget="statusbar"
                     options="{'clickable': '1'}"
                 />
-                <field name="state" invisible="1" />
```

**Comment by hailangvn** [comment]
Updated: as state is `fields.Char`, widget will not work with [this change](https://github.com/OCA/product-attribute/pull/968#issuecomment-997087916).

File: product_tier_validation/models/product_template.py [hunk 1]
```diff
@@ -8,3 +8,34 @@ class ProductTemplate(models.Model):
...
+        self.ensure_one()
+        if "product_state_id" in vals:
+            new_state = self.env["product.state"].browse(vals["product_state_id"]).code
+            return getattr(
+                self, self._state_field
+            ) in self._state_from and new_state not in (
+                self._state_to + [self._cancel_state]
+            )
+        else:
+            return super()._check_tier_state_transition(vals)
```
Comment (hailangvn) [inline product_tier_validation/models/product_template.py:41]: With the improvement at https://github.com/hailangvn/server-ux/pull/1, we can rewrite this like below.
```python
    def _get_state_from(self, vals):
        return self.env["product.state"].browse(vals["product_state_id"]).code
```

File: product_tier_validation/models/product_template.py [hunk 2]
```diff
@@ -8,3 +8,34 @@ class ProductTemplate(models.Model):
     _name = "product.template"
     _inherit = ["product.template", "tier.validation"]
     _tier_validation_manual_config = False
+
```
Comment (hailangvn) [inline product_tier_validation/models/product_template.py:11]: @simahawk, I found where `confirmed` is used.
@rousseldenis, Should we add this?
```python
    _state_to = ["sellable"]
```

File: product_tier_validation/models/product_template.py (hunk 1 above)
Comment (simahawk) [reply product_tier_validation/models/product_template.py:41]: wouldn't be better something like:

```py
[...]
state_code = fields.Char(related="product_state_id.code")`

_state_field = "state_code"
_state_from = ["draft"]
_state_to = ["sellable"]
[...]
```

and we don't override anything?

File: product_tier_validation/models/product_template.py (hunk 2 above)
Comment (rousseldenis) [reply product_tier_validation/models/product_template.py:11]: @hailangvn @simahawk Yes indeed. I first did it in my project to confirm flow.

As 'confirmed' state comes from base_tier_validation, we need to override it here

**Pull request closed without merging**